Merge kubeconfigs using clientcmd.ModifyConfig

vyasgun opened the pull request:
**Fixes:** Issue #3602 

## Solution/Idea

Use `clientcmd.ModifyConfig` to merge custom config with a primary config (global kubeconfig in default case)

## Proposed changes

Changed the function call to ModifyConfig in `start`

## Testing

What is the _bottom-line_ functionality that needs testing? Describe in pseudo-code or in English. Use verifiable statements that tie your changes to existing functionality.

1. `start` should merge the custom kubeconfig to the global kubeconfig file

praveenkumar commented on .gitignore:
```diff
@@ -11,3 +11,4 @@
 /tmp-embed/
 /RPMS/
 /crc
+.idea
```
This should be part of your local git global config instead here.

gbraad replied:
However, some files we can add to gitignore to prevent them littering the repo.

vyasgun replied:
I find keeping this in `.gitignore` is helpful for developing on Goland.

cfergeau replied:
You can add `.idea` to `~/.config/git/ignore`, it's applied to all git repositories, this way you don't need to manually update .gitignore in each project you contribute to.

cfergeau replied:
You marked this conversation as "resolved", not sure if this is because the `~/.config/git/ignore` suggestion works for you, or because you have a strong opinion for having the `.idea` file in crc's `.gitignore`?

praveenkumar commented on pkg/crc/machine/kubeconfig.go:
```diff
@@ -242,33 +242,42 @@ func contains(arr []string, str string) bool {
-	globalConfigPath, kubeConf1, err := getGlobalKubeConfig()
-	if err != nil {
-		return err
+func mergeKubeConfigFile(kubeConfigFile, globalConfigFile string) error {
```
Is this function signature only changed to have test around it?

vyasgun replied:
yes, I changed it because it was being used in just one place and so it can be tested more easily

praveenkumar commented on pkg/crc/machine/kubeconfig_test.go:
```diff
@@ -1,11 +1,14 @@
 package machine
 
 import (
+	"io/ioutil"
```
This is deprecated, use `os` package

gbraad commented on pkg/crc/machine/kubeconfig.go:
```diff
@@ -242,33 +242,42 @@ func contains(arr []string, str string) bool {
+	configAccess.GlobalFile = globalConfigFile
+	configAccess.LoadingRules.ExplicitPath = globalConfigFile
+
+	//fmt.Printf("GUNJAN: %v\n%v\n%v\n", configAccess.GetLoadingPrecedence(), kubeConf1, kubeConf2)
```
Please do not commit debug code ;-)

vyasgun replied:
My bad 😓 Removed it

cfergeau commented on pkg/crc/machine/kubeconfig.go:
```diff
@@ -242,33 +242,42 @@ func contains(arr []string, str string) bool {
-	return clientcmd.WriteToFile(*mergedConfig, globalConfigPath)
+	configAccess := clientcmd.NewDefaultPathOptions()
+	configAccess.GlobalFile = globalConfigFile
+	configAccess.LoadingRules.ExplicitPath = globalConfigFile
```
Looking at https://github.com/kubernetes/client-go/blob/86d49e7265f07676cb39f342595a858b032112de/tools/clientcmd/config.go#L145-L158 , the defaults for `GlobalFile` and `LoadingRules` look useful when `globalKubeConfigFile` is `""`? In other words, maybe this should be changed to:
```
if globalConfigFile != "" {
    configAccess.GlobalFile = globalConfigFile
    configAccess.LoadingRules.ExplicitPath = globalConfigFile
}
```
? Or is this also required when modifying the global kubeconfig file?

vyasgun replied:
I just updated the code again. I found that the `.LocationOfOrigin` field was being given priority in choosing the output file so I changed it to the global kubeconfig file.

cfergeau commented on pkg/crc/machine/kubeconfig.go:
```diff
@@ -242,33 +242,42 @@ func contains(arr []string, str string) bool {
 	for name, context := range kubeConf2.Contexts {
-		mergedConfig.Contexts[name] = context
+		kubeConf1.Contexts[name] = context
 	}
```
This function ends with `return clientcmd.ModifyConfig(configAccess, *kubeConf1, true)` which is documented as:
```
// ModifyConfig takes a Config object, iterates through Clusters, AuthInfos, and Contexts, uses the LocationOfOrigin if specified or uses the default destination file to write the results into.
```

I was hoping that using `ModifyConfig` would allow us to remove these iterations over `Clusters`, `AuthInfos` and `Contexts`?

vyasgun replied:
I couldn't find a way to do it without these iterations

cfergeau commented:
Wondering if `mergeKubeConfigFile` could become as simple as something like this?

```
func mergeKubeConfigFile(kubeConfigFile, globalConfigFile string) error {
	configAccess := clientcmd.NewDefaultPathOptions()
	if globalConfigFile != "" {
		configAccess.GlobalFile = globalConfigFile
		configAccess.LoadingRules.ExplicitPath = globalConfigFile
	}

	kubeConf2, err := clientcmd.LoadFromFile(kubeConfigFile)
	if err != nil {
		return err
	}

	return clientcmd.ModifyConfig(configAccess, *kubeConf2, true)
}
```

praveenkumar commented:
/ok-to-test

cfergeau commented:
> Wondering if `mergeKubeConfigFile` could become as simple as something like this?
> 
> ```
> func mergeKubeConfigFile(kubeConfigFile, globalConfigFile string) error {
> 	configAccess := clientcmd.NewDefaultPathOptions()
> 	if globalConfigFile != "" {
> 		configAccess.GlobalFile = globalConfigFile
> 		configAccess.LoadingRules.ExplicitPath = globalConfigFile
> 	}
> 
> 	kubeConf2, err := clientcmd.LoadFromFile(kubeConfigFile)
> 	if err != nil {
> 		return err
> 	}
> 
> 	return clientcmd.ModifyConfig(configAccess, *kubeConf2, true)
> }
> ```

`ModifyConfig` + `LoadFromFile` behaves oddly, `LoadFromFile` will set `LocationOfOrigin` on all clusters/authinfos/... to the file being loaded. And `ModifyConfig` will write the changes to the file referenced by LocationOfOrigin, which is not what we want :-/

vyasgun commented:
@cfergeau thanks for the review! I couldn't find a way to override the preference for locationOfOrigin field. What can we do next?

cfergeau commented on pkg/crc/machine/kubeconfig.go:
```diff
@@ -242,33 +242,48 @@ func contains(arr []string, str string) bool {
+		globalConfigFile = getGlobalKubeConfigPath()
+	}
+
+	kubeConf1, err = clientcmd.LoadFromFile(globalConfigFile)
```
I'd recommend giving these `kubeConf1` / `kubeConf2` variables more explicit names, to avoid potential confusion. `kubeConf1` could be named `globalKubeConf` for example, it is more obvious what it represents. Or just `destKubeConf`/`srcKubeConf`

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -568,7 +568,7 @@ func (client *client) Start(ctx context.Context, startConfig types.StartConfig)
 		}
 		logging.Info("Adding microshift context to kubeconfig...")
-		if err := mergeKubeConfigFile(constants.KubeconfigFilePath); err != nil {
+		if err := mergeKubeConfigFile(constants.KubeconfigFilePath, ""); err != nil {
```
For what it's worth, adding a dummy argument to a function in the "main" codepath because it will be used in unit testing is not great for readability. One may wonder "What does "" stand for?" while reading the code.

You can achieve the same with an intermediate function 
```
func mergeToGlobalKubeConfigFile(kubeConfigFile string) error {
	return mergeKubeConfigFile(kubeConfigFile, getGlobalKubeConfigPath())
}
```
And remove the special handling of `""` in `mergeKubeConfigFile`.

vyasgun replied:
Thanks, I'll make this change.

cfergeau commented on pkg/crc/machine/kubeconfig_test.go:
```diff
@@ -59,3 +61,75 @@ func TestUpdateUserCaAndKeyToKubeconfig(t *testing.T) {
+	return path, nil
+}
+
+func Test_mergeKubeConfigFile(t *testing.T) {
```
The `TestMergeKubeConfigFile` naming is more common in go tests

vyasgun replied:
I think that's the convention for exported functions

cfergeau reviewed (commented):
> thanks for the review! I couldn't find a way to override the preference for locationOfOrigin field. What can we do next?

Yeah, not really clear what to do :-/ I was hoping using `clientcmd.ModifyConfig` would allow us to remove the need to iterate over the `Contexts`/`AuthInfos`/.. as it's not really useful for crc to know about such details of the kubeconfig content.

It's apparently not possible because `ModifyConfig` does not have the behaviour we want/expected/hoped for, and using it makes `mergeKubeConfigFile` more complex.
On the other hand, this PR adds a unit test which is very good to have! 
Maybe we should change this PR to add the unit test, but keep the current implementation of `mergeKubeConfigFile` as the attempt to use `clientcmd.ModifyConfig` did not turn out as expected?

gbraad commented:
There is a valid point of consideration that this is actually more complex. This is not the fault of the author...

@praveenkumar you think this is an improvement from the perspective of maintenance?

Regardless, the testcase is good to include...

vyasgun commented:
As per the discussions and reviews, I think we can revert the changes made to use `clientcmd.ModifyConfig` as it is not working the way we were expecting it to and just keep the testcase.

anjannath commented on pkg/crc/machine/kubeconfig_test.go:
```diff
@@ -59,3 +61,75 @@ func TestUpdateUserCaAndKeyToKubeconfig(t *testing.T) {
+		for key := range config.Contexts {
+			if _, ok := mergedConfig.Contexts[key]; !ok {
+				t.Errorf("key %v not found", key)
+			}
```
you can also use  `assert.Contains` (https://pkg.go.dev/github.com/stretchr/testify/assert#Contains) in place of the `if` checks
```go
if _, ok := mergedConfig.Contexts[key]; !ok {
	t.Errorf("key %v not found", key)
}
```
this would become:
```go
assert.Contains(t, mergedConfig.Contexts, key, "expected context not found")
```

anjannath commented on pkg/crc/machine/kubeconfig_test.go:
```diff
@@ -59,3 +61,75 @@ func TestUpdateUserCaAndKeyToKubeconfig(t *testing.T) {
+	mergedConfig, err := clientcmd.LoadFromFile(primaryConfigPath)
+	if err != nil {
+		t.Fatalf("failed to load merged kubeconfig: %v", err)
+	}
```
can use `assert.NoError` for the error checks, https://pkg.go.dev/github.com/stretchr/testify/assert#NoError

cfergeau commented on .gitignore:
```diff
@@ -11,3 +11,4 @@
 /tmp-embed/
 /RPMS/
 /crc
+.idea
```
Can we remove this `.idea` change? :)

vyasgun committed:
kubeconfig: Use names more explicit than kubeconf{1,2}

This improves readability, no need to always try to remember if
'kubeconf1' is the global kubeconfig file, or the file we want to merge
in it

vyasgun committed:
Added unit test for mergeConfigFile function

Signed-off-by: Gunjan Vyas <vyasgun20@gmail.com>

vyasgun committed:
kubeconfig: Simplify mergeKubeConfigFile

An intermediate `mergedConfig` is not required, we can directly modify
the global kubeconfig.

cfergeau reviewed (commented):
I find the current set of changes a bit hard to review, as they are doing more than just adding the test case :-/
I've split them in https://github.com/cfergeau/crc/commits/pr/mergeconfig in 3 commits each doing one logical thing.
The first one renames kubeConf1/kubeConf2, the second one simplifies the function a bit (if I understood what it's intended to do), and the last one adds the test case. 
With 3 commits instead of 1, it's easier to be sure we understand what each change is doing, and that there is not an unexpected change which should not be there.

cfergeau commented on pkg/crc/machine/kubeconfig_test.go:
```diff
@@ -59,3 +61,63 @@ func TestUpdateUserCaAndKeyToKubeconfig(t *testing.T) {
 }
+
+func createTempKubeConfig(config *api.Config) (string, error) {
+	tempFile, err := os.CreateTemp("", "kubeconfig-")
```
https://pkg.go.dev/testing#T.TempDir is convenient in tests, no need to check for errors and all its content is automatically cleaned up after the test is done.

cfergeau commented:
/lgtm
/approve
